swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1304 @@
 """
-Particle generation code.
+Generate SPH initial conditions for SPH simulations iteratively for a given density function following Arth et al.
+2019 (https://arxiv.org/abs/1907.11250).
 
-TBD
 """
+import numpy as np
+import unyt
+from math import erf
+from typing import Union
+import warnings
+
+from .IC_kernel import get_kernel_data
+from swiftsimio.optional_packages import KDTree, TREE_AVAILABLE
+from swiftsimio import Writer
+
```

Comment:
This is a massive improvement and I really like the new object-oriented design. I now, unfortunately, have some nit-picky comments :p.

Thanks again for all of your hard work on this!